gen.go

```diff
@@ -855,3 +872,165 @@ func GenTupleEncodersForType(inpkg string, i interface{}, w io.Writer) error {
+		}); err != nil {
+			return err
+		}
+
+		f.Name = "t." + f.Name
+
+		switch f.Type.Kind() {
+		case reflect.String:
+			if err := emitCborMarshalStringField(w, f); err != nil {
+				return err
+			}
+		case reflect.Struct:
+			if err := emitCborMarshalStructField(w, f); err != nil {
+				return err
+			}
+		case reflect.Uint64:
+			if err := emitCborMarshalUint64Field(w, f); err != nil {
+				return err
+			}
+		case reflect.Uint8:
+			if err := emitCborMarshalUint8Field(w, f); err != nil {
+				return err
+			}
+		case reflect.Slice:
+			if err := emitCborMarshalSliceField(w, f); err != nil {
+				return err
+			}
+		case reflect.Bool:
+			if err := emitCborMarshalBoolField(w, f); err != nil {
+				return err
+			}
+		case reflect.Map:
+			if err := emitCborMarshalMapField(w, f); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("field %q of %q has unsupported kind %q", f.Name, gti.Name, f.Type.Kind())
+		}
+	}
+
+	fmt.Fprintf(w, "\treturn nil\n}\n\n")
+	return nil
+}
+
+func emitCborUnmarshalStructMap(w io.Writer, gti *GenTypeInfo) error {
+	err := doTemplate(w, gti, `
+func (t *{{ .Name}}) UnmarshalCBOR(r io.Reader) error {
+	br := cbg.GetPeeker(r)
+
+	maj, extra, err := cbg.CborReadHeader(br)
+	if err != nil {
+		return err
+	}
+	if maj != cbg.MajMap {
+		return fmt.Errorf("cbor input should be of type map")
+	}
+
+	if extra != {{ len .Fields }} {
+		return fmt.Errorf("cbor input had wrong number of fields")
+	}
+
+	var name string
+
+`)
+	if err != nil {
+		return err
+	}
+
+	for _, f := range gti.Fields {
+		fmt.Fprintf(w, "\t// t.%s (%s) (%s)\n", f.Name, f.Type, f.Type.Kind())
+
+		if err := emitCborUnmarshalStringField(w, Field{Name: "name"}); err != nil {
+			return err
+		}
+
+		err := doTemplate(w, f, `
+		if name != "{{ .Name }}" {
+			return fmt.Errorf("expected struct map entry %s to be {{ .Name }}", name)
+		}
+`)
+		if err != nil {
+			return err
+		}
+
+		f.Name = "t." + f.Name
+
+		switch f.Type.Kind() {
+		case reflect.String:
+			if err := emitCborUnmarshalStringField(w, f); err != nil {
+				return err
+			}
+		case reflect.Struct:
+			if err := emitCborUnmarshalStructField(w, f); err != nil {
+				return err
+			}
+		case reflect.Uint64:
+			if err := emitCborUnmarshalUint64Field(w, f); err != nil {
+				return err
+			}
+		case reflect.Uint8:
+			if err := emitCborUnmarshalUint8Field(w, f); err != nil {
+				return err
+			}
+		case reflect.Slice:
+			if err := emitCborUnmarshalSliceField(w, f); err != nil {
+				return err
+			}
+		case reflect.Bool:
+			if err := emitCborUnmarshalBoolField(w, f); err != nil {
+				return err
+			}
+		case reflect.Map:
+			if err := emitCborUnmarshalMapField(w, f); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("field %q of %q has unsupported kind %q", f.Name, gti.Name, f.Type.Kind())
+		}
+	}
+
+	fmt.Fprintf(w, "\treturn nil\n}\n\n")
+
+	return nil
+}
+
+// Generates 'tuple representation' cbor encoders for the given type
```

Comment:
comment copypasta out of date?  this is map repr, not tuple repr, yes?

Reply:
Yes